compiler/src/dmd/dcast.d

```diff
@@ -704,7 +704,7 @@ MATCH implicitConvTo(Expression e, Type t)
                     return MATCH.nomatch;
                 m = MATCH.constant;
             }
-            if (e.hexString && tn.isintegral && (tn.size == e.sz || (!e.committed && (e.len % tn.size) == 0)))
+            if (e.type != t && e.hexString && tn.isintegral && (tn.size == e.sz || (!e.committed && (e.len % tn.size) == 0)))
```

Comment:
This could also be expressed as a separate condition at the beginning of this function:

```
if (e.type == t)
    return MATCH.exact;
```

However, I wanted it to be localized to the actual code that introduced the regression. Arguably, this condition should be first executed by the implicitConvTo function, however, I believe that when it was implemented the expectation was that this would be checked at the call site before actually calling implicitConvTo.